prusti-viper/src/encoder/procedure_encoder.rs

```diff
@@ -1130,160 +1138,262 @@ impl<'p, 'v: 'p, 'tcx: 'v> ProcedureEncoder<'p, 'v, 'tcx> {
-                            encoded_lhs,
-                            ty,
-                            location,
-                            stmt.source_info.span,
-                        )?
-                    }
-                    &mir::Rvalue::Len(ref place) => {
-                        self.encode_assign_sequence_len(
-                            encoded_lhs,
-                            place,
-                            ty,
-                            location,
-                        )?
-                    }
-                    &mir::Rvalue::Repeat(ref operand, ref times) => {
-                        self.encode_assign_array_repeat_initializer(
-                            encoded_lhs,
-                            operand,
-                            times,
-                            ty,
-                            location,
-                        )?
-                    }
-                    ref rhs => {
-                        unimplemented!("encoding of '{:?}'", rhs);
-                    }
+                //
+                // Array access on the LHS should always be mutable (idx is always calculated
+                // before, and just a separate local variable here)
+                let (lhs_place_encoding, ty, _) = self.mir_encoder.encode_place(lhs).unwrap();
+                if let PlaceEncoding::ArrayAccess { box base, index, rust_array_ty, .. } = lhs_place_encoding {
+                    // Current stmt is of the form `arr[idx] = val`. This does not have an expiring
+                    // temporary variable, so we encode it differently from indexing into an array.
+                    self.encode_array_direct_assign(
+                        base,
+                        index,
+                        rust_array_ty,
+                        rhs,
+                        location,
+                    )?
+                } else {
+                    let (encoded_lhs, pre_stmts) = self.postprocess_place_encoding(lhs_place_encoding, ArrayAccessKind::Mutable(None, location))
+                        .with_span(span)?;
+                    stmts.extend(pre_stmts);
+                    self.encode_assign(
+                        encoded_lhs,
+                        rhs,
+                        ty,
+                        location,
+                    )?
                 }
             }
-
             ref x => unimplemented!("{:?}", x),
         };
         stmts.extend(encoding_stmts);
-        Ok(stmts
+        Ok(self.set_stmts_default_pos(stmts, stmt.source_info.span))
+
+    }
+
+    fn set_stmts_default_pos(&self, stmts: Vec<vir::Stmt>, default_pos_span: Span) -> Vec<vir::Stmt> {
+        let expr_pos = self
+            .encoder
+            .error_manager()
+            .register(default_pos_span, ErrorCtxt::GenericExpression);
+        let stmt_pos = self
+            .encoder
+            .error_manager()
+            .register(default_pos_span, ErrorCtxt::GenericStatement);
```

Comment:
Does this not register a lot of errors in the error manager that cannot ever be errors?

Reply:
not sure, i mostly just moved the code into the separate function.